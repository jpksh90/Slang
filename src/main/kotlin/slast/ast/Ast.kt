package slast.ast

import SimpleLangLexer
import SimpleLangParser
import org.antlr.v4.runtime.*

sealed class SlastNode {
    open fun<T> accept(visitor: ASTVisitor<T?>): T? {
        return when (this) {
            is LetStmt -> visitor.visitLetStmt(this)
            is AssignStmt -> visitor.visitAssignStmt(this)
            is FunPureStmt -> visitor.visitFunPureStmt(this)
            is FunImpureStmt -> visitor.visitFunImpureStmt(this)
            is WhileStmt -> visitor.visitWhileStmt(this)
            is PrintStmt -> visitor.visitPrintStmt(this)
            is IfStmt -> visitor.visitIfStmt(this)
            is ExprStmt -> visitor.visitExprStmt(this)
            is ReturnStmt -> visitor.visitReturnStmt(this)
            is IntExpr -> visitor.visitIntExpr(this)
            is BoolExpr -> visitor.visitBoolExpr(this)
            is VarExpr -> visitor.visitVarExpr(this)
            is ReadInputExpr -> visitor.visitReadInputExpr(this)
            is FuncCallExpr -> visitor.visitFuncCallExpr(this)
            is BinaryExpr -> visitor.visitBinaryExpr(this)
            is IfExpr -> visitor.visitIfExpr(this)
            is ParenExpr -> visitor.visitParenExpr(this)
            is Program -> visitor.visitProgram(this)
            is BlockStmt -> visitor.visitBlockStmt(this)
            is NoneValue -> visitor.visitNoneValue(this)
            is Record -> visitor.visitRecord(this)
            is StringExpr -> visitor.visitStringExpr(this)
            is DerefExpr -> visitor.visitDerefExpr(this)
            is RefExpr -> visitor.visitRefExpr(this)
            is DerefStmt -> visitor.visitDerefStmt(this)
            is FieldAccess -> visitor.visitFieldAccessExpr(this)
        }
    }
}

data class Program(val stmt: List<Stmt>) : SlastNode() {
    fun collectFunctionDeclarations() : List<String> {
        val pureFunctions = stmt.filterIsInstance<FunPureStmt>().map { it.name }
        val impureFunctions = stmt.filterIsInstance<FunImpureStmt>().map { it.name }
        return pureFunctions + impureFunctions
    }
}

// Statements
sealed class Stmt : SlastNode()
data class LetStmt(val name: String, val expr: Expr) : Stmt() {
    override fun <T> accept(visitor: ASTVisitor<T?>): T? {
        return expr.accept(visitor)
    }
}

data class AssignStmt(val lhs: Expr, val expr: Expr) : Stmt() {
    override fun <T> accept(visitor: ASTVisitor<T?>): T? {
        return expr.accept(visitor)
    }
}

data class FunPureStmt(val name: String, val params: List<String>, val body: Expr) : Stmt() {
    override fun <T> accept(visitor: ASTVisitor<T?>): T? {
        return body.accept(visitor)
    }
}

data class FunImpureStmt(val name: String, val params: List<String>, val body: BlockStmt) : Stmt() {
    override fun <T> accept(visitor: ASTVisitor<T?>): T? {
        return body.accept(visitor)
    }
}
data class WhileStmt(val condition: Expr, val body: BlockStmt) : Stmt() {
    override fun <T> accept(visitor: ASTVisitor<T?>): T? {
        return body.accept(visitor)
    }
}

data class PrintStmt(val args: List<Expr>) : Stmt() {
    override fun <T> accept(visitor: ASTVisitor<T?>): T? {
        return null
    }
}
data class IfStmt(val condition: Expr, val thenBody: BlockStmt, val elseBody: BlockStmt) : Stmt()
data class ExprStmt(val expr: Expr) : Stmt()
data class ReturnStmt(val expr: Expr) : Stmt()
data class BlockStmt(val stmts: List<Stmt>) : Stmt()
data class DerefStmt(val lhs: Expr, val rhs: Expr) : Stmt()

// Expressions
sealed class Expr : SlastNode()
data class IntExpr(val value: Int) : Expr()
data class BoolExpr(val value: Boolean) : Expr()
data class VarExpr(val name: String) : Expr()
data object ReadInputExpr : Expr()
data class FuncCallExpr(val target: String, val args: List<Expr>) : Expr()
data class BinaryExpr(val left: Expr, val op: String, val right: Expr) : Expr()
data class IfExpr(val condition: Expr, val thenExpr: Expr, val elseExpr: Expr) : Expr()
data class ParenExpr(val expr: Expr) : Expr()
data object NoneValue : Expr()
data class Record(val expression: List<Pair<String, Expr>>) : Expr()
data class StringExpr(val value: String) : Expr()
data class RefExpr(val expr: Expr) : Expr()
data class DerefExpr(val expr: Expr) : Expr()
data class FieldAccess(val lhs: Expr, val rhs: Expr) : Expr()

fun SlastNode.prettyPrint(tabStop: Int = 0): String {
    val indent = "  ".repeat(tabStop)
    return when (this) {
        is BinaryExpr -> "${left.prettyPrint()} ${op} ${right.prettyPrint()}"
        is BoolExpr -> value.toString()
        is FuncCallExpr -> "$target(${args.joinToString(", ") { it.prettyPrint() }})"
        is IfExpr -> "if (${condition.prettyPrint()}) then ${thenExpr.prettyPrint()} else ${elseExpr.prettyPrint()}"
        is IntExpr -> value.toString()
        is ParenExpr -> "(${expr.prettyPrint()})"
        ReadInputExpr -> "readInput()"
        is VarExpr -> name
        is Program -> stmt.joinToString("\n") { it.prettyPrint() }
        is AssignStmt -> "$indent${lhs.prettyPrint()} = ${expr.prettyPrint()};"
        is BlockStmt -> "$indent{\n" + stmts.joinToString("\n") { it.prettyPrint(tabStop + 1) } + "\n$indent}"
        is ExprStmt -> "$indent${expr.prettyPrint()};"
        is FunImpureStmt -> "$indent fun $name(${params.joinToString(", ")}) {\n" + body.prettyPrint(tabStop + 1) + "\n$indent}"
        is NoneValue -> "None"
        is FunPureStmt -> "$indent fun $name(${params.joinToString(", ")}) => ${body.prettyPrint()}"
        is IfStmt -> "$indent if (${condition.prettyPrint()}) {\n" + thenBody.prettyPrint(tabStop + 1) + "\n$indent} else {\n" + elseBody.prettyPrint(tabStop + 1) + "\n$indent}"
        is LetStmt -> "$indent let $name = ${expr.prettyPrint()};"
        is PrintStmt -> "$indent print(${args.joinToString(", ") { it.prettyPrint() }});"
        is ReturnStmt -> "$indent return ${expr.prettyPrint()};"
        is WhileStmt -> "$indent while (${condition.prettyPrint()}) {\n" + body.prettyPrint(tabStop + 1) + "\n$indent}"
        is Record -> "$indent{\n" + expression.joinToString("\n") { "$indent  ${it.first} : ${it.second.prettyPrint()}" } + "\n$indent}"
        is StringExpr -> "\"$value\""
        is DerefExpr -> "deref(${expr.prettyPrint()})"
        is RefExpr -> "ref(${expr.prettyPrint()})"
        is DerefStmt -> "$indent deref(${lhs.prettyPrint()}) = ${rhs.prettyPrint()};"
        is FieldAccess -> "${lhs.prettyPrint()}.${rhs}"
    }
}
//fun ASTNode.isFunctionDeclaration(): Boolean {
//    return when (this) {
//        is FunPureStmt -> true
//        is FunImpureStmt -> true
//        else -> false
//    }
//}


fun main() {
    val inputCode = """
        fun power(base, exp) {
            if (base > exp) {
                base = base + 1;
                return base;
            } else {
                return exp;
            }
        }

        fun topper(x) => x+1;

        let x = true;
    """.trimIndent()
    val inputStream = ANTLRInputStream(inputCode)
    val lexer = SimpleLangLexer(inputStream)
    val tokens = CommonTokenStream(lexer)
    val parser = SimpleLangParser(tokens)

    val parseTree = parser.prog()
    val astBuilder = ASTBuilder()
    val ast = astBuilder.visit(parseTree) as Program
//    print(ast.collectFunctionDeclarations())
    println(ast.prettyPrint())
//    print(prettyPrintAST(ast))
//
//    val collectFunctions = ASTVisitor<List<String>> {
//
//    }
}


//fun main() {
//     val inputCode = """
//        fun power(base, exp) {
//            if (base > exp) {
//                base = base + 1;
//                return base;
//            } else {
//                return exp;
//            }
//        }
//
//        fun topper(x) => x+1;
//
//        let x = true;
//    """.trimIndent()
//    val inputStream = ANTLRInputStream(inputCode)
//    val lexer = SimpleLangLexer(inputStream)
//    val tokens = CommonTokenStream(lexer)
//    val parser = SimpleLangParser(tokens)
//
//    val parseTree = parser.prog()
//    val astBuilder = ASTBuilder()
//    val ast = astBuilder.visit(parseTree) as Program
//
//    val generator = SimpleLangToKotlin()
//    println(generator.generate(ast))
//}


