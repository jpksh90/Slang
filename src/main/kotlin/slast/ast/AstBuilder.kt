package slast.ast

import SimpleLangBaseVisitor
import SimpleLangLexer
import SimpleLangParser
import org.antlr.v4.runtime.ANTLRInputStream
import org.antlr.v4.runtime.CommonTokenStream
import slast.visualizer.CustomErrorListener

class ASTBuilder() : SimpleLangBaseVisitor<SlastNode>() {

    override fun visitLetExpr(ctx: SimpleLangParser.LetExprContext): SlastNode {
        return LetStmt(ctx.ID().text, visit(ctx.expr()) as Expr)
    }

    override fun visitAssignExpr(ctx: SimpleLangParser.AssignExprContext): SlastNode {
        return AssignStmt(visit(ctx.lhs()) as Expr, visit(ctx.expr()) as Expr)
    }

    override fun visitFunPure(ctx: SimpleLangParser.FunPureContext): SlastNode {
        val params = ctx.paramList()?.ID()?.map { it.text } ?: emptyList()
        return FunPureStmt(ctx.ID().text, params, visit(ctx.expr()) as Expr)
    }

    override fun visitFunImpure(ctx: SimpleLangParser.FunImpureContext): SlastNode {
        val params = ctx.paramList()?.ID()?.map { it.text } ?: emptyList()
        val body = ctx.stmt().map { visit(it) as Stmt }
        return FunImpureStmt(ctx.ID().text, params, BlockStmt(body))
    }

    override fun visitWhileStmt(ctx: SimpleLangParser.WhileStmtContext): SlastNode {
        val condition = visit(ctx.expr()) as Expr
        val body = BlockStmt(ctx.blockStmt().stmt().map { visit(it) as Stmt })
        return WhileStmt(condition, body)
    }

    override fun visitPrintStmt(ctx: SimpleLangParser.PrintStmtContext): SlastNode {
        val args = ctx.argList()?.expr()?.map { visit(it) as Expr } ?: emptyList()
        return PrintStmt(args)
    }

    override fun visitIfThenElseStmt(ctx: SimpleLangParser.IfThenElseStmtContext): SlastNode {
        val condition = visit(ctx.expr()) as Expr
        val thenBody = BlockStmt(ctx.blockStmt(0).stmt().map { visit(it) as Stmt })
        val elseBody = BlockStmt(ctx.blockStmt(1).stmt().map { visit(it) as Stmt })
        return IfStmt(condition, thenBody, elseBody)
    }

    override fun visitExprStmt(ctx: SimpleLangParser.ExprStmtContext): SlastNode {
        return ExprStmt(visit(ctx.expr()) as Expr)
    }

    override fun visitReturnStmt(ctx: SimpleLangParser.ReturnStmtContext): SlastNode {
        return ReturnStmt(visit(ctx.expr()) as Expr)
    }

    override fun visitIntExpr(ctx: SimpleLangParser.IntExprContext): SlastNode {
        return NumberLiteral(ctx.NUMBER().text.toDouble())
    }

    override fun visitBoolExpr(ctx: SimpleLangParser.BoolExprContext): SlastNode {
        return BoolLiteral(ctx.BOOL().text.toBoolean())
    }

    override fun visitVarExpr(ctx: SimpleLangParser.VarExprContext): SlastNode {
        return VarExpr(ctx.ID().text)
    }

    override fun visitReadInputExpr(ctx: SimpleLangParser.ReadInputExprContext): SlastNode {
        return ReadInputExpr
    }

    override fun visitFuncCallExpr(ctx: SimpleLangParser.FuncCallExprContext): SlastNode {
        val args = ctx.argList()?.expr()?.map { visit(it) as Expr } ?: emptyList()
        val target = visit(ctx.expr()) as Expr
        if (target is VarExpr) {
            return FuncCallExpr(target.name, args)
        } else {
            return FuncCallExpr("unresolved", args)
        }
    }

    override fun visitArithmeticExpr(ctx: SimpleLangParser.ArithmeticExprContext): SlastNode {
        return BinaryExpr(visit(ctx.expr(0)) as Expr, ctx.op.text, visit(ctx.expr(1)) as Expr)
    }

    override fun visitComparisonExpr(ctx: SimpleLangParser.ComparisonExprContext): SlastNode {
        return BinaryExpr(visit(ctx.expr(0)) as Expr, ctx.op.text, visit(ctx.expr(1)) as Expr)
    }

    override fun visitBooleanExpr(ctx: SimpleLangParser.BooleanExprContext): SlastNode {
        return BinaryExpr(visit(ctx.expr(0)) as Expr, ctx.op.text, visit(ctx.expr(1)) as Expr)
    }

    override fun visitIfExpr(ctx: SimpleLangParser.IfExprContext): SlastNode {
        return IfExpr(
            visit(ctx.expr(0)) as Expr,
            visit(ctx.expr(1)) as Expr,
            visit(ctx.expr(2)) as Expr
        )
    }

    override fun visitParenExpr(ctx: SimpleLangParser.ParenExprContext): SlastNode {
        return ParenExpr(visit(ctx.expr()) as Expr)
    }

    override fun visitForStmt(ctx: SimpleLangParser.ForStmtContext): SlastNode {
        val initialization = AssignStmt(visit(ctx.ID()) as Expr, visit(ctx.expr(0)) as Expr)
        val condition = visit(ctx.expr(1)) as Expr
        val update = visit(ctx.expr(2)) as Stmt
        val body = ctx.blockStmt().stmt().map { visit(it) as Stmt }
        val block = BlockStmt(listOf(initialization, WhileStmt(condition, BlockStmt(body + update))))
        return block
    }

    override fun visitProg(ctx: SimpleLangParser.ProgContext): SlastNode {
        val statements = ctx.stmt().map { visit(it) as Stmt }
        return Program(statements)
    }

    override fun visitNoneValue(ctx: SimpleLangParser.NoneValueContext): SlastNode {
        return NoneValue
    }

    override fun visitRecordExpr(ctx: SimpleLangParser.RecordExprContext): SlastNode {
        val recordIds = ctx.recordElems().ID()
        val recordExprs = ctx.recordElems().expr()
        val recordElementPairs = mutableListOf<Pair<String, Expr>>()
        for (i in 0 until recordIds.size) {
            recordElementPairs.addLast(Pair(recordIds[i].text, visit(recordExprs[i]) as Expr))
        }
        return Record(recordElementPairs)
    }

    override fun visitStringExpr(ctx: SimpleLangParser.StringExprContext): SlastNode {
        return StringLiteral(ctx.STRING().text)
    }

    override fun visitRefExpr(ctx: SimpleLangParser.RefExprContext): SlastNode {
        return RefExpr(visit(ctx.expr()) as Expr)
    }

    override fun visitDerefExpr(ctx: SimpleLangParser.DerefExprContext): SlastNode {
        return DerefExpr(visit(ctx.deref()) as Expr)
    }

    override fun visitDeref(ctx: SimpleLangParser.DerefContext): SlastNode {
        return visit(ctx.expr())
    }

    override fun visitLhs(ctx: SimpleLangParser.LhsContext): SlastNode {
        if (ctx.deref() != null) {
            return DerefExpr(visit(ctx.deref()) as Expr)
        }

        if (ctx.fieldAccess() != null) {
            return visit(ctx.fieldAccess())
        }
        return VarExpr(ctx.ID().text)
    }

    override fun visitPrimaryExprWrapper(ctx: SimpleLangParser.PrimaryExprWrapperContext): SlastNode {
        return visit(ctx.primaryExpr())
    }

    override fun visitFieldAccess(ctx: SimpleLangParser.FieldAccessContext): SlastNode {
        return FieldAccess(visit(ctx.expr()) as Expr, VarExpr(ctx.ID().text))
    }

    override fun visitFieldAccessExpr(ctx: SimpleLangParser.FieldAccessExprContext): SlastNode {
        return FieldAccess(visit(ctx.expr()) as Expr, VarExpr(ctx.ID().text))
    }

    override fun visitDoWhileStmt(ctx: SimpleLangParser.DoWhileStmtContext): SlastNode {
        val body = visit(ctx.blockStmt()) as Stmt
        val condition = visit(ctx.expr())
        return BlockStmt(listOf(body, WhileStmt(condition as Expr, body as BlockStmt)))
    }

    override fun visitBlockStmt(ctx: SimpleLangParser.BlockStmtContext): SlastNode {
            return BlockStmt(ctx.stmt().map { visit(it) as Stmt })
    }

}

fun main(args: Array<String>) {
    val x = """fun foo(a) => a;
        let t = foo;
        t("this is a string");
    """.trimMargin()

    fun parseProgram(input: String): SlastNode {
        val parser = SimpleLangParser(CommonTokenStream(SimpleLangLexer(ANTLRInputStream(input))))

        parser.removeErrorListeners()
        val errorListener = CustomErrorListener()
        parser.addErrorListener(errorListener)

        val parseTree = parser.prog()

        val astBuilder = ASTBuilder()
        val ast = astBuilder.visit(parseTree) as Program
        return ast
    }

    print(parseProgram(x))

}
