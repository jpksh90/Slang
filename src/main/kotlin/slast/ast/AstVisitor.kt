package slast.ast

interface ASTVisitor<T> {
    fun visitLetStmt(stmt: LetStmt): T
    fun visitAssignStmt(stmt: AssignStmt): T
    fun visitFunPureStmt(stmt: FunPureStmt): T
    fun visitFunImpureStmt(stmt: FunImpureStmt): T
    fun visitWhileStmt(stmt: WhileStmt): T
    fun visitPrintStmt(stmt: PrintStmt): T
    fun visitIfStmt(stmt: IfStmt): T
    fun visitExprStmt(stmt: ExprStmt): T
    fun visitReturnStmt(stmt: ReturnStmt): T
    fun visitBlockStmt(stmt: BlockStmt): T

    fun visitIntExpr(expr: NumberLiteral): T
    fun visitBoolExpr(expr: BoolLiteral): T
    fun visitVarExpr(expr: VarExpr): T
    fun visitReadInputExpr(expr: ReadInputExpr): T
    fun visitFuncCallExpr(expr: FuncCallExpr): T
    fun visitBinaryExpr(expr: BinaryExpr): T
    fun visitIfExpr(expr: IfExpr): T
    fun visitParenExpr(expr: ParenExpr): T
    fun visitProgram(program: Program): T
    fun visitNoneValue(expr: NoneValue): T

    fun visitRecord(expr: Record): T
    fun visitStringExpr(expr: StringLiteral): T

    fun visitRefExpr(expr: RefExpr): T
    fun visitDerefExpr(expr: DerefExpr): T
    fun visitDerefStmt(stmt: DerefStmt): T
    fun visitFieldAccessExpr(expr: FieldAccess): T
}

abstract class BaseASTVisitor<T> : ASTVisitor<T> {
    protected open fun defaultVisit(): T {
        throw UnsupportedOperationException("Operation not supported")
    }

    override fun visitLetStmt(stmt: LetStmt): T = defaultVisit()
    override fun visitAssignStmt(stmt: AssignStmt): T = defaultVisit()
    override fun visitFunPureStmt(stmt: FunPureStmt): T = defaultVisit()
    override fun visitFunImpureStmt(stmt: FunImpureStmt): T = defaultVisit()
    override fun visitWhileStmt(stmt: WhileStmt): T = defaultVisit()
    override fun visitPrintStmt(stmt: PrintStmt): T = defaultVisit()
    override fun visitIfStmt(stmt: IfStmt): T = defaultVisit()
    override fun visitExprStmt(stmt: ExprStmt): T = defaultVisit()
    override fun visitReturnStmt(stmt: ReturnStmt): T = defaultVisit()
    override fun visitBlockStmt(stmt: BlockStmt): T = defaultVisit()

    override fun visitIntExpr(expr: NumberLiteral): T = defaultVisit()
    override fun visitBoolExpr(expr: BoolLiteral): T = defaultVisit()
    override fun visitVarExpr(expr: VarExpr): T = defaultVisit()
    override fun visitReadInputExpr(expr: ReadInputExpr): T = defaultVisit()
    override fun visitFuncCallExpr(expr: FuncCallExpr): T = defaultVisit()
    override fun visitBinaryExpr(expr: BinaryExpr): T = defaultVisit()
    override fun visitIfExpr(expr: IfExpr): T = defaultVisit()
    override fun visitParenExpr(expr: ParenExpr): T = defaultVisit()
    override fun visitProgram(program: Program): T = defaultVisit()
    override fun visitNoneValue(expr: NoneValue): T = defaultVisit()

    override fun visitRecord(expr: Record): T = defaultVisit()
    override fun visitStringExpr(expr: StringLiteral): T = defaultVisit()

    override fun visitRefExpr(expr: RefExpr): T = defaultVisit()
    override fun visitDerefExpr(expr: DerefExpr): T = defaultVisit()
    override fun visitDerefStmt(stmt: DerefStmt): T = defaultVisit()
    override fun visitFieldAccessExpr(expr: FieldAccess): T = defaultVisit()
}



