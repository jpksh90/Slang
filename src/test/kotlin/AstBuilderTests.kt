import org.antlr.v4.runtime.ANTLRInputStream
import org.antlr.v4.runtime.CommonTokenStream
import org.approvaltests.Approvals
import slast.ast.ASTBuilder
import slast.ast.prettyPrint
import java.net.URL
import kotlin.test.Test

class AstBuilderTests {

    fun loadTestProgram(name: String): URL? {
        return AstBuilderTests::class.java.getResource("/$name")
    }

    fun buildAst(input: String): String {
        val lexer = SimpleLangLexer(ANTLRInputStream(input))
        val parser = SimpleLangParser(CommonTokenStream(lexer))
        val tree = parser.prog()
        val astBuilder = ASTBuilder()
        return astBuilder.visit(tree).prettyPrint()
    }

    @Test
    fun testCase1() {
        val program = loadTestProgram("exponent.sl")
        if (program != null) {
            val ast = buildAst(program.readText())
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase2() {
        val program = loadTestProgram("facl.sl")
        if (program != null) {
            val ast = buildAst(program.readText())
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase3() {
        val program = loadTestProgram("gcd.sl")
        if (program != null) {
            val ast = buildAst(program.readText())
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase4() {
        val program = loadTestProgram("minmax.sl")
        if (program != null) {
            val ast = buildAst(program.readText())
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase5() {
        val program = loadTestProgram("disallowed.sl")
        if (program != null) {
            val ast = buildAst(program.readText())
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase6() {
        val program = loadTestProgram("sum_prod.sl")
        if (program != null) {
            val ast = buildAst(program.readText())
            Approvals.verify(ast)
        }
    }
}