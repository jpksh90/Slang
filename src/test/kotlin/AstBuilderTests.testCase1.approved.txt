 fun power(base, exp) {
  {
     let result = 1.0;
     while (exp > 0.0) {
      {
        result = result * base;
        exp = exp - 1.0;
      }
    }
     return result;
  }
}
 let base = readInput();
 let exp = readInput();
 print(power(base, exp));